### PR TITLE
Sprint5

### PR DESCRIPTION
**Pull request by Ashutosh-Tigga-NucleusTeq**

### CODE REVIEW

File: src/main/java/com/blog/portal/controller/CommentPostController.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+@RestController
+@RequestMapping(RequestMappingConst.COMMENT_URL)
+public class CommentPostController {
+
+	/**
+	 * Instance  CommentService.
+	 */
+	@Autowired
+	private CommentPostService commentPostService;
```
Comment (Abhishek-Hasda-nucleusteq) [inline src/main/java/com/blog/portal/controller/CommentPostController.java:31]: follow the proper indentation.

File: src/main/java/com/blog/portal/controller/BlogPostController.java [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+	 * An instance of the {@link BlogPostService} class for handling blog post
+	 * operations.
+	 */
+	@Autowired
+	private BlogPostService blogPostService;
+
+	/**
+	 * An instance of the Logger class for logging.
+	 */
+	private Logger logger = LogManager.getLogger(BlogPostController.class);
```
Comment (Abhishek-Hasda-nucleusteq) [inline src/main/java/com/blog/portal/controller/BlogPostController.java:51]: make it final and the variable should be in capital letter. LOGGER

**Review by Abhishek-Hasda-nucleusteq — CHANGES_REQUESTED** [review]

File: src/main/java/com/blog/portal/serviceimpl/BlogPostServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+	/**
+	 * Method that deal with creation of post API in database.
+	 */
+	@Override
+	public ApiResponse createPost(@Valid PostBlogInDto postBlogInDto) {
+
+		ApiResponse reponse = null;
+		User user = blogUserRepo.findById(postBlogInDto.getUserId())
+				.orElseThrow(() -> new ResourceNotFoundException(null, null, null));
+		user.setPassword(null);
```
[the remark this answers is not in the document]
Comment (Ashutosh-Tigga-NucleusTeq) [reply src/main/java/com/blog/portal/serviceimpl/BlogPostServiceImpl.java:65]: i am setting null in password bcz it's i don't wanna reveal password to cliend bcz i am sending these data to client side this is not to save data in database

File: src/main/java/com/blog/portal/requestPayload/AuthenticateUserInDto.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+import javax.validation.constraints.Pattern;
+
+/**
+ * This class Represents a Request DTO used for authenticating a user. It
+ * contains the user's email address and password for authentication.
+ *
+ * @author [ Ashutosh Tigga]
+ */
+public class AuthenticateUserInDto {
+	/**
```
Comment (Krishna-NucleusTeq) [inline src/main/java/com/blog/portal/requestPayload/AuthenticateUserInDto.java:14]: UserInDTO

File: src/main/java/com/blog/portal/serviceimpl/BlogServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+			fetchedPost = blogRepository.findByTechCategoryAndStatus(inDto.getTechCategory(), status);
+		}
+		if (inDto.getTitle() != null && inDto.getTechCategory() == null) {
+			fetchedPost = blogRepository.findByTitleContainingIgnoreCaseAndStatus(inDto.getTitle(), status);
+		}
+		if (inDto.getTitle() != null && inDto.getTechCategory() != null) {
+			fetchedPost = blogRepository.findByTitleContainingIgnoreCaseAndStatusAndTechCategory(
+					inDto.getTitle(), status,
+					inDto.getTechCategory());
+		}
```
Comment (Krishna-NucleusTeq) [inline src/main/java/com/blog/portal/serviceimpl/BlogServiceImpl.java:109]: what  if fetchedPost will empty?

File: src/main/java/com/blog/portal/serviceimpl/BlogServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+		if (inDto.getTitle() != null && inDto.getTechnologyCategory() == null) {
+			fetchedListOfPost = blogRepository.findByTitleContainingIgnoreCaseAndStatus(
+					inDto.getTitle(),
+					status);
+		}
+		if (inDto.getTitle() != null && inDto.getTechnologyCategory() != null) {
+			fetchedListOfPost = blogRepository.findByTitleContainingIgnoreCaseAndStatusAndTechCategory(
+					inDto.getTitle(),
+					status,
+					inDto.getTechnologyCategory());
```
Comment (Krishna-NucleusTeq) [inline src/main/java/com/blog/portal/serviceimpl/BlogServiceImpl.java:250]: same as above

File: src/main/java/com/blog/portal/services/CommentService.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+public interface CommentService {
+
+	/**
+	 * THis method to do comment on post.
+	 *
+	 * @param inDto
+	 * @return CommentOutDto
+	 */
+	ApiResponse doCommentOnBlog(CommentOnBlogInDto inDto);
+
```
Comment (Krishna-NucleusTeq) [inline src/main/java/com/blog/portal/services/CommentService.java:27]: method name should be postComment

File: src/main/java/com/blog/portal/services/ReactionService.java [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+@Service
+public interface ReactionService {
+
+	/**
+	 * perform like or dislike on post.
+	 *
+	 * @param inDto
+	 * @return LikeOrDislikeOutDto.
+	 */
+	ReactionOnBlogOutDto doReactOnBlog(ReactOnBlogInDto inDto);
```
Comment (Krishna-NucleusTeq) [inline src/main/java/com/blog/portal/services/ReactionService.java:23]: BlogActivity

File: src/main/java/com/blog/portal/services/UserService.java [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+	ApiResponse createUser(@Valid RegisterUserInDto user);
+
+	/**
+	 * method that helps us to Authenticate user.
+	 *
+	 * @param inDto
+	 * @return AuthenticateOutDto.
+	 */
+	AuthenticateUserOutDto authenticateUser(@Valid AuthenticateUserInDto inDto);
+
```
Comment (Krishna-NucleusTeq) [inline src/main/java/com/blog/portal/services/UserService.java:32]: UserOutDTO

File: src/main/java/com/blog/portal/util/BlogConst.java [hunk 10]
```diff
@@ -0,0 +1,12 @@
+package com.blog.portal.util;
+
+public class BlogConst {
+	/**
```
Comment (Krishna-NucleusTeq) [inline src/main/java/com/blog/portal/util/BlogConst.java:4]: Constants

File: src/main/java/com/blog/portal/util/BlogConst.java [hunk 11]
```diff
@@ -0,0 +1,12 @@
...
+	 * Constant class name.
+	 */
+	public static final String CLASS_NAME = "Blog";
+	/**
+	 * Constant Field name.
+	 */
+	public static final String FIELD_POST_ID = "postId";
+}
```
Comment (Krishna-NucleusTeq) [inline src/main/java/com/blog/portal/util/BlogConst.java:12]: POST_ID

File: src/main/java/com/blog/portal/util/ExceptionMessage.java [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+public class ExceptionMessage {
+	/**
+	 * message: Email already exists.
+	 */
+	public static final String EMAIL_ALREADY_EXISTS = "Email Already registered";
+	/**
+	 * message: the user is not authenticated.
+	 */
+	public static final String UNAUTHORIZED_USER = "The user is not authenticated";
+}
```
Comment (Krishna-NucleusTeq) [inline src/main/java/com/blog/portal/util/ExceptionMessage.java:12]: message should be Unauthorized user

File: src/main/java/com/blog/portal/util/RequestMappingConst.java [hunk 13]
```diff
@@ -0,0 +1,24 @@
+package com.blog.portal.util;
+
+public class RequestMappingConst {
+	/**
```
Comment (Krishna-NucleusTeq) [inline src/main/java/com/blog/portal/util/RequestMappingConst.java:4]: RestPathConstants

File: src/main/java/com/blog/portal/util/ResponseMessage.java [hunk 14]
```diff
@@ -0,0 +1,92 @@
+package com.blog.portal.util;
+
+/**
+ * This class contains response messages that are returned by the backend. It's
+ * important to ensure that these messages are clear and informative.
+ */
+public class ResponseMessage {
+	/**
```
Comment (Krishna-NucleusTeq) [inline src/main/java/com/blog/portal/util/ResponseMessage.java:8]: class should be final

File: src/main/java/com/blog/portal/util/ExceptionMessage.java [hunk 15]
```diff
@@ -0,0 +1,12 @@
+package com.blog.portal.util;
+
+public class ExceptionMessage {
+	/**
```
Comment (Krishna-NucleusTeq) [inline src/main/java/com/blog/portal/util/ExceptionMessage.java:4]: ErrorConstants

File: src/main/java/com/blog/portal/util/UserConst.java
Comment (Krishna-NucleusTeq) [inline src/main/java/com/blog/portal/util/UserConst.java:1]: Keep all the constants in one file, error messages in ErrorConstants, and url should be in restpathConstants

File: src/test/java/com/blog/portal/serviceimpl/UserServiceImplTest.java [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+
+        ApiResponse response = userService.createUser(userDto);
+
+        assertNotNull(response);
+        assertTrue(response.isSuccess());
+        assertEquals(ResponseMessage.USER_REGISTER_SUCCESS, response.getMessage());
+    }
+
+    @Test
+    public void testCreateUser_UserAlreadyExists() {
```
Comment (Krishna-NucleusTeq) [inline src/test/java/com/blog/portal/serviceimpl/UserServiceImplTest.java:61]: testUserAlreadyExists

File: src/test/java/com/blog/portal/serviceimpl/UserServiceImplTest.java [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+        User user = RegisterUserMapper.inDtoToUser(userDto);
+
+        when(userRepo.findByEmail(user.getEmail())).thenReturn(Optional.of(user));
+
+        assertThrows(UserAlreadyExistsException.class, () -> userService.createUser(userDto));
+    }
+
+    @Test
+    public void testAuthenticateUser_Success() {
+        AuthenticateUserInDto authDto = new AuthenticateUserInDto();
```
Comment (Krishna-NucleusTeq) [inline src/test/java/com/blog/portal/serviceimpl/UserServiceImplTest.java:76]: testUserAuthenticatedSuccessfully

File: src/test/java/com/blog/portal/serviceimpl/UserServiceImplTest.java [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+        User user = new User();
+        user.setEmail(authDto.getEmail());
+        user.setPassword(authDto.getPassword());
+
+        when(userRepo.findByEmail(authDto.getEmail())).thenReturn(Optional.of(user));
+
+        AuthenticateUserOutDto outDto = userService.authenticateUser(authDto);
+
+        assertNotNull(outDto);
+        assertEquals(authDto.getEmail(), outDto.getEmail());
```
Comment (Krishna-NucleusTeq) [inline src/test/java/com/blog/portal/serviceimpl/UserServiceImplTest.java:89]: no need to check notNull, check whole object. expected and actual

File: src/test/java/com/blog/portal/serviceimpl/UserServiceImplTest.java [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+
+        when(userRepo.findByEmail(user.getEmail())).thenReturn(Optional.empty());
+        when(userRepo.save(any(User.class))).thenReturn(user);
+
+        ApiResponse response = userService.createUser(userDto);
+
+        assertNotNull(response);
+        assertTrue(response.isSuccess());
+        assertEquals(ResponseMessage.USER_REGISTER_SUCCESS, response.getMessage());
+    }
```
Comment (Krishna-NucleusTeq) [inline src/test/java/com/blog/portal/serviceimpl/UserServiceImplTest.java:58]: no need to check notNull, check whole object. expected and actual

File: src/test/java/com/blog/portal/serviceimpl/ReactionServiceImplTest.java [hunk 20]
```diff
@@ -0,0 +1,141 @@
...
+        when(blogRepository.findById("postId")).thenReturn(Optional.of(blog));
+        when(blogRepository.save(blog)).thenReturn(blog);
+        when(likeAndDislikeRepo.save(any(Reaction.class))).thenAnswer(invocation -> invocation.getArgument(0));
+
+        ReactionOnBlogOutDto result = reactionService.doReactOnBlog(inDto);
+
+        assertNotNull(result);
+        assertEquals(React.LIKE, result.getType());
+        assertTrue(blog.getLikedBy().contains("userId"));
+        assertFalse(blog.getDisLikedBy().contains("userId"));
```
Comment (Krishna-NucleusTeq) [inline src/test/java/com/blog/portal/serviceimpl/ReactionServiceImplTest.java:58]: no need to check notNull and fields individually, check whole object. expected and actual

**Review by Krishna-NucleusTeq — CHANGES_REQUESTED** [review]

File: src/main/java/com/blog/portal/repository/BlogRepository.java [hunk 21]
```diff
@@ -0,0 +1,146 @@
...
+          String title, BlogStatus status, TechnologyCategory techCategory);
+
+  /**
+   * Query for fetching posts by title.
+   * @param title
+   * @param status
+   * @return List<Blog>
+   */
+  List<Blog>  findByTitleContainingIgnoreCaseAndStatus(String title, BlogStatus status);
+
```
[the remark this answers is not in the document]
Comment (Ashutosh-Tigga-NucleusTeq) [reply src/main/java/com/blog/portal/repository/BlogRepository.java:36]: It is for filtering Blog by mentioned  criteria it's like custom query

**Pull request merged**